File: 2130/maximum-twin-sum-of-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        vector<int> a;

        ListNode * next = head;
        int N = 0;
        while(next!=nullptr){
            a.push_back(next->val);
            next = next->next;
            N++;
        }
        int ans = 0;//maximum twin sum
        for(int i=0; i<N/2; i++){
            int temp = a[i] + a[N-1-i];
            if(temp>ans) ans = temp;
        }
        return ans;
    }
};

File: 2130/maximum-twin-sum-of-a-linked-list.py
# LeetCode 2130. Maximum Twin Sum of a Linked List
# 問「頭尾相加」的組合，加起來「最大值」有多大？
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        ans = 0
        # 把 Linked List 先複製成 array 即可
        a = []
        head2 = head
        while head != None:
            a.append(head.val)
            head = head.next
        # 複製好後，再用「倒過來的 for 迴圈，逐一檢查頭尾
        for i in range(len(a)-1, -1, -1):
            ans = max(ans, head2.val + a[i])
            head2 = head2.next
        return ans
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
